p2p/src/swarm/mod.rs

```diff
@@ -96,314 +86,395 @@ where
+        if let Some(Some(channel)) = self.pending.remove(&address) {
+            channel.send(Err(error)).map_err(|_| P2pError::ChannelClosed)?;
+        }
 
-                if self.peers.len() == MAX_ACTIVE_CONNECTIONS {
-                    log::warn!("maximum number of connections reached, close new connection with peer {:?}", peer_id);
-                    // TODO: save peer information for later?
-                    // TODO: i.e., consider this a peer discovery event?
-                    return self.handle.disconnect(peer_id).await;
-                }
+        self.peerdb.report_outbound_failure(address);
+        Ok(())
+    }
 
-                if peer_info.magic_bytes != *self.config.magic_bytes() {
-                    log::error!(
-                        "peer {:?} is in different network, ours {:?}, theirs {:?}",
-                        peer_id,
-                        peer_info.magic_bytes,
-                        self.config.chain_type()
-                    );
-                    return Err(P2pError::ProtocolError(ProtocolError::DifferentNetwork(
-                        *self.config.magic_bytes(),
-                        peer_info.magic_bytes,
-                    )));
-                }
+    /// Attempt to establish an outbound connection
+    ///
+    /// This function doesn't block on the call but sends a command to the
+    /// networking backend which then reports at some point in the future
+    /// whether the connection failed or succeeded.
+    async fn connect(&mut self, address: T::Address) -> crate::Result<()> {
+        // TODO: verify that the peer is not already part of our swarm (needs peerdb)
+        ensure!(
+            !self.pending.contains_key(&address),
+            P2pError::PeerError(PeerError::Pending(address.to_string()))
+        );
+        ensure!(
+            self.validate_address(&address),
+            P2pError::PeerError(PeerError::BannedAddress(address.to_string())),
+        );
 
-                // TODO: check supported protocols
-                // TODO: check version
+        self.handle.connect(address).await
+    }
 
-                self.peers.insert(peer_id, PeerContext { _info: peer_info });
-                self.tx_sync
-                    .send(event::SyncControlEvent::Connected(peer_id))
-                    .await
-                    .map_err(P2pError::from)
-            }
-            net::types::ConnectivityEvent::ConnectionAccepted { peer_info } => {
-                let peer_id = peer_info.peer_id;
-                log::debug!("outbound connection accepted by peer {:?}", peer_id);
+    /// Maintain the swarm state
+    ///
+    /// `PeerManager::heartbeat()` is called every time a network/control event is received
+    /// or the heartbeat timer of the event loop expires. In other words, the swarm state
+    /// is checked and updated at least once every 30 seconds. In high-traffic scenarios the
+    /// update interval is clamped to a sensible lower bound.
```

Comment:
Maybe briefly explain how it's lowered (what part does it)

Reply:
If you're talking about the lower bound then my thought was just to clamp it to like 5 seconds or something. Each time `heartbeat()` is called it checks when it was last called (when it executed the whole function) and if it's more than 5 seconds (or whatever the lower bound is configured to be), it exits early.

Reply:
I meant in the comment mention where that mechanism is done :)

Reply:
Done